File: JAVA/LINKEDLIST/doubly_LL.java
public class doubly_LL {
    class node{
        public int data;
        public node next;
        public node prev;

        node(int data){
            this.data=data;
            this.next=null;
            this.prev=null;
        }
    } 
    public node head=null;
    public node tail=null;

    public void displaylist(){
        node curr=head;
        if(head==null){
            System.out.println("This given list is empty");
            return;
        }
        System.out.println("data is");
        while(curr!=null){
            System.out.println(curr.data+" ");
            curr=curr.next;
        }
        System.out.println();
    }
    public void reverse(){
        while(tail.next!=null){
            tail=tail.next;
        }
        while(head.prev!=null){
            System.out.print(tail.data+"->");
            tail=tail.prev;
        }
        System.out.print("null");
    }
    public void insert(int data){
        node node1=new node(data);
        if(head==null){
            head=tail=node1;
            tail.next=null;
            head.prev=null;
        }else{
            tail.next=node1;
            node1.prev=tail;
            node1.next=null;
            
        }
    }

    public static void main(String[] args) {
       doubly_LL list=new doubly_LL();
       list.insert(1);
       list.insert(2);
       list.insert(3);

       list.insert(4);
       list.displaylist();

    }
}

File: C++/DoublyLL.cpp
// Question: Remove duplicates from an sorted doubly linked list (DLL)
#include<iostream>
using namespace std;

class node{  //defining structure of DLL
public:
    node * prev; //condition for DLL
    node *  next;
    int data;

    node(int data){  //constructor

        this->data=data;
        prev=nullptr;
        next=nullptr;
    }

}; 

class DoublyLL{
public:
    node * head;

    DoublyLL(){   //constructor
        head=nullptr;
    }

};

void displayLL(node * head){
    if(head == nullptr) {    //if a list is empty
        cout << "NULL";
        return;
    }

    node * temp = head;    //saving head node in temp node

    while(temp)
    {

        cout << temp->data << "<->";
        temp = temp->next; 
    }

    cout << "NULL";
}

void removeDuplicates(node* &head)
{
    if(head==NULL)
    {
        cout<<"Empty Doubly Linked List!!";
        return;
    }
    
    if(head -> next == NULL) 
    {
        cout << "Single Node in doubly Linked List" << endl;
        return ;
    }
    
    node* curr = head;
    
    while(curr!=NULL)
    {
        if(curr->next!=NULL && (curr->data == curr->next->data))
        {
            node* temp = curr->next;
            curr->next = temp->next;
            temp->next = NULL;    //to avoid memory leak
            delete temp;
            
        }
        
        else
        {
             curr = curr->next; 
        }
    }
}

int main(){

    DoublyLL dll; //initializing 
    dll.head = new node(1);
    dll.head->next = new node(1);
    dll.head->next->prev = dll.head;
    dll.head->next->next = new node(2);
    dll.head->next->next->prev = dll.head->next;
    dll.head->next->next->next = new node(2);
    dll.head->next->next->next->prev = dll.head->next->next;

    cout << "Original List: "<<endl;
    displayLL(dll.head);

    cout<<endl;


    removeDuplicates(dll.head); 

    cout << "List after removing duplicates: "<<endl;;
    displayLL(dll.head);
 
 return 0;
}

